Fix tab navigation inconsistency between QuantityStepper and QuantitySelector components

amitraj2203 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR addresses accessibility and UX inconsistencies between the `QuantityStepper` component used in the `AddToCartsWithOptions` block and the `QuantitySelector` component used in the `Mini-Cart` Drawer block. While these components are structurally different, they provide a similar user interface and should behave consistently.

- Specifically, this PR includes:
  - Reordered the tab index in `QuantityStepper` so that:
  - Focus moves first to the input field, then the minus (-) button, and finally the plus (+) button - matching the `QuantitySelector` component behavior.
  - ~Applied correct disabled, unfocusable, and unclickable behavior to the minus (-) button when quantity equals the minimum allowed value.~

Part Of: #58011 

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

#### Blockified: 
https://github.com/user-attachments/assets/35cd31e6-a400-43f5-bcd5-6eb0e920ef03

#### Non-blockified: 
https://github.com/user-attachments/assets/4d07b5b4-4bf9-460c-899a-9097c86981d1

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install and activate the `WooCommerce Beta Tester plugin`
2. Go to *WooCommerce Admin Test Helper > Features* and enable `experimental-blocks` and `blockified-add-to-cart`
3. Go to *Appearance > Editor > Templates > Single Product*, select the `Add to Cart with Options` block and update it to the blockified one
4. View a product in the frontend
5. Use the Tab key to navigate through the `QuantityStepper` component
6. Ensure tab order is: **input field → plus (-) button → minus (+) button**.
7. ~Set quantity to the minimum allowed value (e.g., 1 or custom minimum via filter).~
8. ~Confirm minus (-) button is:~
   * ~Not focusable with tab~
   * ~Not clickable~

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

amitraj2203 committed:
refactor: QuantityStepper buttons to render after the input element and adjust CSS order for layout

amitraj2203 committed:
fix: Correct order of quantity stepper buttons in AddToCartForm and Utils

amitraj2203 committed:
feat: Make QuantityStepper minus button disabled when the min value matches with current value

amitraj2203 committed:
fix: Remove unnecessary blank line in style.scss

amitraj2203 committed:
feat: Add changelog file

Aljullu reviewed (commented):
Thanks for working on this, @amitraj2203!

>   - Reordered the tab index in `QuantityStepper` so that:
>  - Focus moves first to the input field, then the minus (-) button, and finally the plus (+) button - matching the `QuantitySelector` component behavior.

This works great and code changes look good too. :100: 

>   - Applied correct disabled, unfocusable, and unclickable behavior to the minus (-) button when quantity equals the minimum allowed value.

This is also working as described, but we should try to implement this using the [Interactivity API](https://developer.wordpress.org/block-editor/reference-guides/interactivity-api/).

With it, it shouldn't be needed to query HTML elements directly, instead you could add a directive like `data-wp-bind--disabled="state.allowsDecrease"` to the minus button or something along those lines. Then `allowsDecrease` could be a state in the [stores](https://github.com/woocommerce/woocommerce/blob/8ddea4bf169c071167726833a7cc8c32653ecd84/plugins/woocommerce/client/blocks/assets/js/blocks/add-to-cart-with-options/frontend.ts#L120) that checks whether `value - step` is higher or equal than `minValue`. Similarly, we should implement the same for the plus button, as it's technically possible with custom code or extensions that products might have max quantities too.

Let me know how this sounds or if you have any doubts. :slightly_smiling_face:

amitraj2203 committed:
feat: Update quantity stepper buttons to use iAPI enabling/disabling

amitraj2203 committed:
feat: Update quantity stepper buttons to use iAPI enabling/disabling for AddToCartForm block

amitraj2203 committed:
fix: lint error

amitraj2203 committed:
fix: Refactor getInputElementFromEvent to improve readability

amitraj2203 committed:
fix: remove unnecessary change

amitraj2203 commented:
Thanks for the review @Aljullu, I've now updated it accordingly. Please let me know if everything looks good now or if any further changes are needed.

Aljullu reviewed (commented):
Great job! Code looks good and it's working mostly fine, I just found one small glitch in grouped products:

If the first product's quantity is increased, then the minus button of the other products is not disabled correctly:

![image](https://github.com/user-attachments/assets/226ac07b-6a2b-4697-ad16-331dab1919b5)

Same the other way around, if the first product can't be reduced, all minus buttons get disabled:

![image](https://github.com/user-attachments/assets/10fe1e1a-0d10-4c58-b906-3eacb092c76f)

Can you reproduce as well? I guess that's because this selector is too generic and always selects the first input:

```JS
document.querySelector(
	'.wc-block-components-quantity-selector__input'
)
```

Maybe we could use the input ID or something similar to properly select the correct element? (If you go this way, I would advise rebasing or merging `trunk` first, as we modified the input IDs in https://github.com/woocommerce/woocommerce/pull/57927).

amitraj2203 commented:
Thanks for catching this @Aljullu! You're absolutely right about the issue. 

The problem is that `const { productId } = getContext< Context >()` was giving the ID of the grouped product (e.g., 35 for "Logo Collection") not the IDs of individual products inside that grouped product (e.g., 15 for "Hoodie", 16 for "T-Shirt").

We are using a **single shared store (state.allowsIncrease and state.allowsDecrease)** for *all* quantity steppers. However, each quantity input corresponds to a different product row (e.g. ID 15, 16), but **the store logic doesn't distinguish between them**. It checks all inputs and returns false if *any* of them are at their min/max.

As a result, **both** buttons become disabled if **any one** input hits its min/max, even if others are still adjustable.

To fix this issue, I think we need to:

1. **Make the input selection context-aware**: Instead of using the generic `document.querySelector( '.wc-block-components-quantity-selector__input' )`, we should use the product ID from context to find the specific input: `document.getElementById( quantity_${ productId } )`.

2. **Ensure each product gets its own context**: Each individual product in a grouped product should have its own interactive context with the correct `productId`.

We can also apply this approach to **all quantity inputs**, not only for grouped products. For simple/variable products, `const { productId } = getContext< Context >()` will return the correct productId of that specific product, and for grouped products, each individual product will have its own context with its own productId.

This way, each quantity stepper will only check its own input element and be independent of others.

Let me know if this approach correct, or do you have any other suggestions for handling this?

Aljullu commented:
Thanks for investigating this, @amitraj2203! Actually, in #58002 we might transform the `quantity` state into an object that matches product IDs with their quantities. So I think it might be a good idea to pause this PR, and update it once https://github.com/woocommerce/woocommerce/pull/58002 is merged into `trunk`. I think the work from that PR is exactly what is missing here, and I would like to avoid ending up implementing the same thing twice or differently. Do you think that's the right approach?

amitraj2203 commented:
Sure thing @Aljullu, I'll wait for it to be merged into `trunk` and then continue with this.

Aljullu commented:
@amitraj2203 heads-up that https://github.com/woocommerce/woocommerce/pull/58002 has been merged. So you should be able to rebase or merge `trunk` into this branch. :slightly_smiling_face:

amitraj2203 committed:
Merge branch 'trunk' of github.com:amitraj2203/woocommerce into fix/quantity-stepper-accessibility-and-tab-order

amitraj2203 committed:
feat: Enhance quantity input interactivity by adding child product ID support

amitraj2203 committed:
feat: Update quantity handling for grouped products to support child product IDs

amitraj2203 commented:
Hi @Aljullu,

Sorry for the delay and thank you for your patience. I was a bit tied up with other tasks recently, but I’ve now addressed the [issue](https://github.com/woocommerce/woocommerce/pull/58307#pullrequestreview-2887579946) and fixed the bug. Let me know if there’s anything else you'd like me to update or adjust.

Thanks!

https://github.com/user-attachments/assets/cb3a4a4b-95b9-4619-8328-de68e365e265

amitraj2203 committed:
refactor: Simplify quantity input element selection for grouped products

Aljullu commented on plugins/woocommerce/client/blocks/assets/js/blocks/add-to-cart-with-options/frontend.ts:
```diff
@@ -165,6 +166,70 @@ const addToCartWithOptionsStore = store(
+				const parsedMinValue = parseInt( inputElement.min, 10 );
+				const parsedStep = parseInt( inputElement.step, 10 );
+
+				const minValue = isNaN( parsedMinValue ) ? 1 : parsedMinValue;
```
In grouped products children, we need to allow setting a quantity of `0`. It's being fixed in https://github.com/woocommerce/woocommerce/pull/58997, but in order to avoid conflicts between the two PRs, I would set the default `minValue` to `0` when `childProductId` is defined.

amitraj2203 replied:
Thanks I've done it that way now.

amitraj2203 replied:
@Aljullu I've included the fix for the _Add to Cart with Options block (the non-blockified version)_ in this PR only. Additionally, there was a bug where changing the quantity through the input field didn't activate the decrease stepper button — I've included a fix for that too.

Aljullu reviewed (commented):
Thanks for updating this PR again, @amitraj2203! :raised_hands: 

This is mostly working fine, I noticed however the same issue as before in the existing _Add to Cart with Options_ block (the non-blockified one): when I increase or decrease the quantity of a grouped product child, all the quantity selector buttons update according to the last modified one:

https://github.com/user-attachments/assets/eb7868ca-c018-4f06-b10f-06018814d5ca

:point_up_2: notice that after setting _T-Shirt_ to `0`, the button to decrease the quantity of _Hoodie with Logo_ is disabled.

Can you reproduce as well? Btw, I'm fine if we fix that in this PR or if we split this improvement in two separate PRs, one for each block, as you prefer. :slightly_smiling_face:

amitraj2203 committed:
feat: Allow quantity of 0 for grouped product children in quantity validation

amitraj2203 committed:
fix: Issue of quantity selector buttons updating according to the last modified product's quantity

amitraj2203 committed:
fix: Issue of descrease stepper not becoming active on change of quantity through input

amitraj2203 committed:
revert last change for blockified add-to-cart-with-options block

amitraj2203 committed:
Merge branch 'trunk' of github.com:amitraj2203/woocommerce into fix/quantity-stepper-accessibility-and-tab-order

amitraj2203 committed:
fix: Adds missing arguments of function

amitraj2203 committed:
feat: Fix the input class selecter issue

Aljullu commented on plugins/woocommerce/client/blocks/assets/js/blocks/add-to-cart-with-options/frontend.ts:
```diff
@@ -169,6 +170,74 @@ const addToCartWithOptionsStore = store(
+						? `.wc-block-components-quantity-selector__input[name="quantity[${ childProductId }]"]`
+						: `.wc-block-components-quantity-selector__input[name="quantity"]`;
+
+				const inputElement = document.querySelector(
```
Using `document.querySelector()` here might be problematic because there might be the same product displayed as a top-level product and as a grouped product child. Ie:

```
Single Product: Logo collection
--> Beanie
--> T-Shirt
Single Product: Beanie
```

The first `Beanie` will have a min value of `0`, while the second one will have a min value of `1`. So `document.querySelector( ... ).min` will return 1 or 0 depending on the order of the blocks.

This will require a significant refactor, so we could handle that in a separate PR, but I think we should move away from getting the `min`, `step` and `max` values from the HTML input and instead pass it as the Interactivity API context from the backend. What do you think?

amitraj2203 replied:
@Aljullu I've reverted the change that _disables the minus (-) button when the quantity equals the minimum allowed value_.

This PR now includes only the following:

> - Reordered the tab index in `QuantityStepper` so that:
>  - Focus moves first to the input field, then to the minus (-) button, and finally to the plus (+) button — matching the behavior of the `QuantitySelector` component.

---

I’ll raise a new PR to address the issue of **disabling the minus (-) button when the quantity equals the minimum allowed value**, and I’ll incorporate the three comments you provided in this PR as part of that work.

Aljullu commented on plugins/woocommerce/client/blocks/assets/js/blocks/add-to-cart-with-options/frontend.ts:
```diff
@@ -169,6 +170,74 @@ const addToCartWithOptionsStore = store(
+					return false;
+				}
+
+				const parsedMaxValue = parseInt( inputElement.max, 10 );
+				const parsedStep = parseInt( inputElement.step, 10 );
+
+				const maxValue = isNaN( parsedMaxValue )
+					? undefined
+					: parsedMaxValue;
+				const step = isNaN( parsedStep ) ? 1 : parsedStep;
+
+				return (
+					maxValue === undefined || currentQuantity + step <= maxValue
+				);
```
Small nit, but I think using an early return when `parsedMaxValue` is NaN would make it easier to read:

```suggestion
				const parsedMaxValue = parseInt( inputElement.max, 10 );

				if ( isNaN( parsedMaxValue ) ) {
					return true;
				}
				
				const parsedStep = parseInt( inputElement.step, 10 );
				const step = isNaN( parsedStep ) ? 1 : parsedStep;

				return currentQuantity + step <= maxValue;
```

This is quite subjective, so I'm fine leaving it as-is if you prefer. :slightly_smiling_face: In case you end up changing this, it can be changed in the `add-to-cart-form/frontend.ts` file as well.

Aljullu commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-elements/add-to-cart-form/frontend.ts:
```diff
@@ -32,34 +39,118 @@ const getInputData = ( event: HTMLElementEvent< HTMLButtonElement > ) => {
+				childProductId,
+			} = inputData;
 			const newValue = currentValue + step;
 
```
I wonder if we should save a variable like this at the beginning of the functions:

```TS
			const productId = childProductId || context.productId;
			if ( ! productId ) {
				return false;
			} 
```

If you do so, you will be able to avoid the `if { ... } else { ... }` and won't have to add `as number` every time you use `context.productId`.

Aljullu reviewed (commented):
Thanks for all the updates, @amitraj2203! This looks promising, but I left a few small comments inline. I'm a bit concerned about the usage of `document.querySelector()` to get the input values, as it might return any `input` on the page: from other blocks, classic templates, etc. And I feel that won't be completely reliable.

I know this PR is growing quite a lot, so I'm fine if you prefer splitting it. I think at least these two items:

> * Reordered the tab index in QuantityStepper so that:
> * Focus moves first to the input field, then the minus (-) button, and finally the plus (+) button - matching the QuantitySelector component behavior.

Are stable now, so they could be merged separately from the other one, which seems to be the most complex one:

> * Applied correct disabled, unfocusable, and unclickable behavior to the minus (-) button when quantity equals the minimum allowed value.

amitraj2203 committed:
revert the changes of making the minus (-) button disabled when quantity equals the minimum allowed value.

Aljullu reviewed (approved):
Thanks for splitting this in two separate PRs, this way we can merge these improvements already. :slightly_smiling_face: 

I tested, and the focus order is as described now.

https://github.com/user-attachments/assets/358cb570-ca6b-4854-84cb-0f2b856e4294

<details>
<summary>Testing environment</summary>

* WordPress 6.8.1 local dev environment
* Plugins: WooCommerce Beta Tester
* Theme: Woo Starter Theme
* Store creation: June 2025
* Store size: small
</details>

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypes/AddToCartForm.php:
```diff
@@ -75,14 +75,14 @@ protected function enqueue_data( array $attributes = [] ) {
+		// Replacement string to add button AFTER the matched <input> element.
 		/* translators: %s refers to the item name in the cart. */
-		$minus_button = '<button aria-label="' . esc_attr( sprintf( __( 'Reduce quantity of %s', 'woocommerce' ), $product_name ) ) . '"type="button" data-wp-on--click="actions.removeQuantity" class="wc-block-components-quantity-selector__button wc-block-components-quantity-selector__button--minus">-</button>$1';
+		$minus_button = '$1<button aria-label="' . esc_attr( sprintf( __( 'Reduce quantity of %s', 'woocommerce' ), $product_name ) ) . '"type="button" data-wp-on--click="actions.removeQuantity" class="wc-block-components-quantity-selector__button wc-block-components-quantity-selector__button--minus">-</button>';
```
Sorry, I missed that. There is a space missing here:

```suggestion
		$minus_button = '$1<button aria-label="' . esc_attr( sprintf( __( 'Reduce quantity of %s', 'woocommerce' ), $product_name ) ) . '" type="button" data-wp-on--click="actions.removeQuantity" class="wc-block-components-quantity-selector__button wc-block-components-quantity-selector__button--minus">-</button>';
```

amitraj2203 replied:
Fixed the missing space. Also I just changed the description of this PR to `Part Of`. Hoping it shouldn't closes the issue 😅

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptions/Utils.php:
```diff
@@ -21,14 +21,14 @@ class Utils {
+		// Replacement string to add button AFTER the matched <input> element.
 		/* translators: %s refers to the item name in the cart. */
-		$minus_button = '<button aria-label="' . esc_attr( sprintf( __( 'Reduce quantity of %s', 'woocommerce' ), $product_name ) ) . '" type="button" data-wp-on--click="actions.decreaseQuantity" class="wc-block-components-quantity-selector__button wc-block-components-quantity-selector__button--minus">-</button>$1';
+		$minus_button = '$1<button aria-label="' . esc_attr( sprintf( __( 'Reduce quantity of %s', 'woocommerce' ), $product_name ) ) . '"type="button" data-wp-on--click="actions.decreaseQuantity" class="wc-block-components-quantity-selector__button wc-block-components-quantity-selector__button--minus">-</button>';
```
And here:

```suggestion
		$minus_button = '$1<button aria-label="' . esc_attr( sprintf( __( 'Reduce quantity of %s', 'woocommerce' ), $product_name ) ) . '" type="button" data-wp-on--click="actions.decreaseQuantity" class="wc-block-components-quantity-selector__button wc-block-components-quantity-selector__button--minus">-</button>';
```

amitraj2203 committed:
Update plugins/woocommerce/src/Blocks/BlockTypes/AddToCartForm.php

Co-authored-by: Albert Juhé Lluveras <aljullu@gmail.com>

amitraj2203 committed:
Update plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptions/Utils.php

Co-authored-by: Albert Juhé Lluveras <aljullu@gmail.com>

Aljullu reviewed (approved):
Thanks for the updates and the contribution, @amitraj2203! LGTM! :raised_hands:
